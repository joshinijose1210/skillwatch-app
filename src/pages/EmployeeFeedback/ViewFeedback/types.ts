export interface FeedbackDataTypes {
    date: Date;
    isExternalFeedback: boolean;
    externalFeedbackFromEmailId: string;
    feedback: string;
    feedbackToEmployeeId: string;
    feedbackFromEmployeeId: string;
    tagId: number;
    feedbackType: string;
    toEmpName: string;
    fromEmpName: string;
    toRoleName: string;
    fromRoleName: string;
}
