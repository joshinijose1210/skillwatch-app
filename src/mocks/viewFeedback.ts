export const ViewFeedbackDataMockData = {
    key: '',
    pathname: '',
    search: '',
    hash: '',
    state: [
        {
            date: new Date('2024-10-18T00:00:00Z'),
            isExternalFeedback: true,
            externalFeedbackFromEmailId: 'external@gmail.com',
            feedback: 'this is a test feedback for view feedback page testing',
            feedbackToEmployeeId: 'T0123',
            feedbackFromEmployeeId: 'EmployeeId',
            tagId: 12,
            feedbackType: 'Positive',
            toEmpName: 'EmployeeName',
            fromEmpName: 'FromName',
            toRoleName: 'Dev',
            fromRoleName: 'Dev2'
        },
        {
            date: new Date('2024-10-18T00:00:00Z'),
            isExternalFeedback: false,
            externalFeedbackFromEmailId: 'external@gmail.com',
            feedback: 'this is a test feedback for view feedback page testing',
            feedbackToEmployeeId: 'T0123',
            feedbackFromEmployeeId: 'EmployeeId',
            tagId: 12,
            feedbackType: 'Positive',
            toEmpName: 'EmployeeName',
            fromEmpName: 'FromName',
            toRoleName: 'Dev',
            fromRoleName: 'Dev2'
        }
    ]
};
