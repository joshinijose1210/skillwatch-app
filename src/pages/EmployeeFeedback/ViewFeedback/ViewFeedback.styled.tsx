import { Text } from '@medly-components/core';
import styled from 'styled-components';

export const FeedbackLayer = styled.div`
    display: flex;
    flex-direction: column;
    span {
        font-size: ${({ theme }) => theme.contentFontSize};
    }
`;

export const DetailsContainer = styled.div`
    display: flex;
    flex-direction: column;
    max-width: 65%;
    gap: 2rem;
    margin-left: 1rem;
`;

export const StyledHeading = styled(Text)`
    font-size: ${({ theme }) => theme.contentFontSize};
`;
export const StyledHTMLText = styled(Text)`
    p {
        word-wrap: break-word;
        margin-block-start: 0;
        font-size: ${({ theme }) => theme.contentFontSize};
    }
`;
export const StyledEmployeeRole = styled(Text)`
    color: ${({ theme }) => theme.colors.grey[500]};
`;
