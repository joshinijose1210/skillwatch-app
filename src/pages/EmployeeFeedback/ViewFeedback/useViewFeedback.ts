import { useLocation } from 'react-router-dom';
import { FeedbackDataTypes } from './types';

export const useViewFeedback = () => {
    const state = useLocation().state as FeedbackDataTypes;
    return {
        state
    };
};
