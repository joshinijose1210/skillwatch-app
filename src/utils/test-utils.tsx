import { ToastContainer } from '@medly-components/core';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import { store as rtkStore, setupStore } from '@slice';
import { RenderOptions, RenderResult, render } from '@testing-library/react';
import { defaultTheme } from '@theme';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import reduxMockStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import { userMockState } from '../mocks/userMockState';
import { initialState } from '../slice';

interface CustomOptions extends RenderOptions {
    initialState?: any;
    initialEntries?: any;
}

interface WithThemeProviderProps {
    children?: React.ReactNode;
}

const mockStore = reduxMockStore(),
    store = mockStore({
        ...initialState,
        user: { ...initialState.user, groups: ['admin'] }
    });

const WithThemeProvider: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <ThemeProvider theme={defaultTheme}>
        <>
            <ToastContainer position="top-end" />
            {children}
        </>
    </ThemeProvider>
);

const WithStore: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <WithThemeProvider>{children}</WithThemeProvider>
    </Provider>
);

const WithRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <MemoryRouter>
        <WithThemeProvider>{children}</WithThemeProvider>
    </MemoryRouter>
);

const WithStoreAndRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <MemoryRouter>
            <WithThemeProvider>{children}</WithThemeProvider>
        </MemoryRouter>
    </Provider>
);

export const RTKRender = (ui: React.ReactElement, { initialState = userMockState, ...renderOptions }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);

    const RtkWrapper: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
        <Provider store={store}>
            <MemoryRouter initialEntries={renderOptions.initialEntries}>
                <WithThemeProvider>{children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
    return { rtkStore, ...render(ui, { wrapper: RtkWrapper, ...renderOptions }) };
};
const customRender = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithThemeProvider, ...options });

export const renderWithStore = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStore, ...options });

export const renderWithRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithRouter, ...options });

export const renderWithStoreAndRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStoreAndRouter, ...options });

export const wrapper = (props: { children: React.ReactNode }, { initialState = userMockState }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);
    return (
        <Provider store={store}>
            <MemoryRouter>
                <WithThemeProvider>{props.children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
};

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render, store };
