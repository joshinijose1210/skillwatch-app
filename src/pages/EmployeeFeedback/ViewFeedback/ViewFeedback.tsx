import { PageContent } from '@components';
import { Text } from '@medly-components/core';
import format from 'date-fns/format';
import DOMPurify from 'dompurify';
import { useViewFeedback } from './useViewFeedback';
import { DetailsContainer, FeedbackLayer, StyledEmployeeRole, StyledHeading, StyledHTMLText } from './ViewFeedback.styled';
import ListHeader from '@components/reusableComponents/ListHeader';

const ViewFeedback = () => {
    const { state } = useViewFeedback();
    return (
        <PageContent>
            <ListHeader title={'View Feedback'} />
            <DetailsContainer>
                <FeedbackLayer>
                    <StyledHeading textWeight="Medium">
                        Date: <Text>{format(new Date(state.date), 'dd/MM/yyyy')}</Text>
                    </StyledHeading>
                </FeedbackLayer>
                <FeedbackLayer>
                    <StyledHeading textVariant="h4" textWeight="Medium">
                        Feedback From:
                    </StyledHeading>
                    <Text>
                        {state.isExternalFeedback
                            ? state.externalFeedbackFromEmailId
                            : `${state.fromEmpName} (${state.feedbackFromEmployeeId})`}
                    </Text>
                    <StyledEmployeeRole>
                        <i>{state.fromRoleName}</i>
                    </StyledEmployeeRole>
                </FeedbackLayer>
                <FeedbackLayer>
                    <StyledHeading textVariant="h4" textWeight="Medium">
                        Feedback About:
                    </StyledHeading>
                    <Text>
                        {state.toEmpName} ( {state.feedbackToEmployeeId} )
                    </Text>
                    <StyledEmployeeRole>
                        <i>{state.toRoleName}</i>
                    </StyledEmployeeRole>
                </FeedbackLayer>
                <FeedbackLayer>
                    <StyledHeading textVariant="h4" textWeight="Medium">
                        Feedback Type:
                    </StyledHeading>
                    <Text>{state.feedbackType}</Text>
                </FeedbackLayer>
                <FeedbackLayer>
                    <StyledHeading textVariant="h4" textWeight="Medium">
                        Feedback:
                    </StyledHeading>
                    <StyledHTMLText dangerouslySetInnerHTML={{ __html: DOMPurify.sanitize(state.feedback) }} />
                </FeedbackLayer>
            </DetailsContainer>
        </PageContent>
    );
};

export default ViewFeedback;
