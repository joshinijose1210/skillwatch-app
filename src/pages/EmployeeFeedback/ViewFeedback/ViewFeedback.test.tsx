import { RTKRender, cleanup, screen } from '@utils/test-utils';
import ViewFeedback from './ViewFeedback';
import * as reactRouterDom from 'react-router-dom';

import { apiInstance } from '@utils';
import MockAdapter from 'axios-mock-adapter';
import { ViewFeedbackDataMockData } from '@mocks/viewFeedback';
const mock = new MockAdapter(apiInstance);

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: jest.fn(),
    useNavigation: mockNavigate
}));

const useLocationMock = jest.spyOn(reactRouterDom, 'useLocation');

afterEach(() => {
    cleanup();
    mock.reset();
});

describe('ViewFeedback component when external feedback is receieved', () => {
    useLocationMock.mockReturnValue({ ...ViewFeedbackDataMockData, state: { ...ViewFeedbackDataMockData.state[0] } });

    it('should render properly', () => {
        const { container } = RTKRender(<ViewFeedback />);
        expect(container).toMatchSnapshot();
    });

    it('show external email', () => {
        RTKRender(<ViewFeedback />);
        expect(screen.getByText('external@gmail.com')).toBeInTheDocument();
    });
});

describe('ViewFeedback component when internal feedback is receieved', () => {
    it("should show from employee's name and id", () => {
        useLocationMock.mockReturnValue({ ...ViewFeedbackDataMockData, state: { ...ViewFeedbackDataMockData.state[1] } });
        RTKRender(<ViewFeedback />);
        expect(screen.getByText('FromName (EmployeeId)')).toBeInTheDocument();
    });
});
