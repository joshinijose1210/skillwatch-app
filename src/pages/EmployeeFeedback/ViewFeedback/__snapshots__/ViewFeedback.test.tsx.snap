// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewFeedback component when external feedback is receieved should render properly 1`] = `
<div>
  <main
    class="sc-1ehavxz-0 dDyhBJ"
  >
    <div
      class="sc-h70dbj-0 fQmYvl row expanded space-between align-center marginBottom-2"
    >
      <div
        class="sc-w8e7z6-2"
      >
        <h3
          class="sc-1ntgl3l-0 kcLBOi sc-w8e7z6-0 ibnmt"
          textalign="initial"
          textcolor="inherit"
          textvariant="h3"
          textweight="Medium"
        >
          View Feedback
        </h3>
      </div>
    </div>
    <div
      class="sc-u1crwe-1 gyJsin"
    >
      <div
        class="sc-u1crwe-0 lgvtAn"
      >
        <span
          class="sc-1ntgl3l-0 gPMgWC sc-u1crwe-2 cnczNT"
          textalign="initial"
          textcolor="inherit"
          textweight="Medium"
        >
          Date: 
          <span
            class="sc-1ntgl3l-0 hGagSU"
            textalign="initial"
            textcolor="inherit"
          >
            18/10/2024
          </span>
        </span>
      </div>
      <div
        class="sc-u1crwe-0 lgvtAn"
      >
        <h4
          class="sc-1ntgl3l-0 jPeJxZ sc-u1crwe-2 cnczNT"
          textalign="initial"
          textcolor="inherit"
          textvariant="h4"
          textweight="Medium"
        >
          Feedback From:
        </h4>
        <span
          class="sc-1ntgl3l-0 hGagSU"
          textalign="initial"
          textcolor="inherit"
        >
          external@gmail.com
        </span>
        <span
          class="sc-1ntgl3l-0 hGagSU sc-u1crwe-4 hOTIgn"
          textalign="initial"
          textcolor="inherit"
        >
          <i>
            Dev2
          </i>
        </span>
      </div>
      <div
        class="sc-u1crwe-0 lgvtAn"
      >
        <h4
          class="sc-1ntgl3l-0 jPeJxZ sc-u1crwe-2 cnczNT"
          textalign="initial"
          textcolor="inherit"
          textvariant="h4"
          textweight="Medium"
        >
          Feedback About:
        </h4>
        <span
          class="sc-1ntgl3l-0 hGagSU"
          textalign="initial"
          textcolor="inherit"
        >
          EmployeeName
           ( 
          T0123
           )
        </span>
        <span
          class="sc-1ntgl3l-0 hGagSU sc-u1crwe-4 hOTIgn"
          textalign="initial"
          textcolor="inherit"
        >
          <i>
            Dev
          </i>
        </span>
      </div>
      <div
        class="sc-u1crwe-0 lgvtAn"
      >
        <h4
          class="sc-1ntgl3l-0 jPeJxZ sc-u1crwe-2 cnczNT"
          textalign="initial"
          textcolor="inherit"
          textvariant="h4"
          textweight="Medium"
        >
          Feedback Type:
        </h4>
        <span
          class="sc-1ntgl3l-0 hGagSU"
          textalign="initial"
          textcolor="inherit"
        >
          Positive
        </span>
      </div>
      <div
        class="sc-u1crwe-0 lgvtAn"
      >
        <h4
          class="sc-1ntgl3l-0 jPeJxZ sc-u1crwe-2 cnczNT"
          textalign="initial"
          textcolor="inherit"
          textvariant="h4"
          textweight="Medium"
        >
          Feedback:
        </h4>
        <span
          class="sc-1ntgl3l-0 hGagSU sc-u1crwe-3 eOrgCc"
          textalign="initial"
          textcolor="inherit"
        >
          this is a test feedback for view feedback page testing
        </span>
      </div>
    </div>
  </main>
</div>
`;
